### DISCUSSION TITLE
AllChem.Compute2DCoords with coordMap produces clashing depictions

### DISCUSSION THREAD

**Discussion by laurenreid1** [rdkit/rdkit#6518] [Q&A]
Hi,

RDKit version: 2020.09.4

I'm using Compute2DCoords with the coordMap argument to align 2 molecule depictions. I have a specific atom map dictionary that contains the atoms I want to ensure have the same coordinates between the 2 compounds.  For the majority of cases, I find that the following procedure gives the desired output:

- generate 2D coordinates for small molecule
- use the atom map to generate a 2D coordinate dictionary for the large molecule to match the desired atoms to the coordinates from the small molecule
- generate 2D coordinates for large molecule using Compute2DCoords with the coordMap argument

However, the following example produces a large molecule depiction where the non-matching atoms (e.g the sidechain leaving the chiral carbon) clashes with the fixed atoms:

```
def draw_molecule(mol, outfile):
    d2d = rdMolDraw2D.MolDraw2DCairo(300, 200)
    d2d.DrawMolecule(mol)
    d2d.FinishDrawing()
    d2d.WriteDrawingText(outfile)

small_smiles = "c1ccc2c(c1)cncc2NC(=O)C3CCSc4c3cc(cc4)Cl"
large_smiles = "c1ccc2c(c1)cncc2NC(=O)C3(CCOc4c3cc(cc4)Cl)CC(=O)N5CCCC5Cn6ccnc6"

## atom map between 2 compounds (generated by proceeding cheminformatics process)
atom_map = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10, 11: 11, 12: 12, 13: 13, 14: 14, 15: 15, 16: 16, 17: 17, 18: 18, 19: 19, 20: 20, 21: 21, 22: 22}

small_mol = AllChem.MolFromSmiles(small_smiles)
large_mol = AllChem.MolFromSmiles(large_smiles)

## generate 2D coordinates for small molecule
AllChem.Compute2DCoords(small_mol)

## get 2D coordinates from small molecule and store in dictionary keyed by corresponding atom id in large molecule
coordDict2D = {}
for large_mol_atom_id, small_mol_atom_id in atom_map.items():
    pt = small_mol.GetConformer().GetAtomPosition(small_mol_atom_id)
    coordDict2D[large_mol_atom_id] = Geometry.Point2D(pt.x, pt.y)
    
## generate 2D coordinates for large molecule with coordMap
AllChem.Compute2DCoords(large_mol, coordMap=coordDict2D)

draw_molecule(large_mol, 'large_mol.png')
draw_molecule(small_mol, 'small_mol.png')
```

where small_mol.png = 
![small_mol](https://github.com/rdkit/rdkit/assets/8626999/2531eb59-852e-4d9e-822d-df848184420e)

and large_mol.png = 
![large_mol](https://github.com/rdkit/rdkit/assets/8626999/f64b7ac8-3ab4-4b1a-b039-de4d30641cc5)

I can see that the atoms where I fixed the coordinates are depicted how I'd expect, but the remaining atoms in the molecule are overlayed on top of them. Is there a way to ensure that the non-fixed atoms are depicted using non-clashing coordinates?

I have tried altering the sampling settings (e.g nSample and nFlipsPerSample) but the resulting depictions always contain clashes. 

Thanks for any help or thoughts on this!

**Comment by ptosco** [comment]
Hi Lauren, this is not really a solution to your problem but rather a workaround.
When it comes to constrained depictions either using a `coordMap` or a fixed scaffold with `GenerateDepictionMatching2DStructure`, in my experience Schr&ouml;dinger's `CoordGen` does a better job and avoids such clashes.
See below - `CoordGen` will even shift the original substituent to accommodate the additional substituent on the chromane ring and adjust the geometry of the attachment points.
```
from rdkit import Chem, Geometry
from rdkit.Chem.Draw import rdDepictor, rdMolDraw2D

rdDepictor.SetPreferCoordGen(False)

small_smiles = "c1ccc2c(c1)cncc2NC(=O)C3CCSc4c3cc(cc4)Cl"
large_smiles = "c1ccc2c(c1)cncc2NC(=O)C3(CCOc4c3cc(cc4)Cl)CC(=O)N5CCCC5Cn6ccnc6"

## atom map between 2 compounds (generated by proceeding cheminformatics process)
atom_map = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10, 11: 11, 12: 12, 13: 13, 14: 14, 15: 15, 16: 16, 17: 17, 18: 18, 19: 19, 20: 20, 21: 21, 22: 22}

small_mol = Chem.MolFromSmiles(small_smiles)
large_mol = Chem.MolFromSmiles(large_smiles)

## generate 2D coordinates for small molecule
rdDepictor.Compute2DCoords(small_mol)
0

## get 2D coordinates from small molecule and store in dictionary keyed by corresponding atom id in large molecule
coordDict2D = {}
for large_mol_atom_id, small_mol_atom_id in atom_map.items():
    pt = small_mol.GetConformer().GetAtomPosition(small_mol_atom_id)
    coordDict2D[large_mol_atom_id] = Geometry.Point2D(pt.x, pt.y)

rdDepictor.SetPreferCoordGen(True)

## generate 2D coordinates for large molecule with coordMap using CoordGen
rdDepictor.Compute2DCoords(large_mol, coordMap=coordDict2D)
0

## normalize the CoordGen depiction to RDKit standard bond length
rdDepictor.NormalizeDepiction(large_mol, canonicalize=0)
1.5

large_mol
```
![image](https://github.com/rdkit/rdkit/assets/5244385/9121bf3d-299b-4fb3-9904-d827eef3ca04)

```
small_mol
```
![image](https://github.com/rdkit/rdkit/assets/5244385/419d6f00-ff97-44c2-83bd-6d98c2a1a7d8)

**Comment by laurenreid1** [comment]
Hi Paolo,

Thanks for the advice. I will try out CoordGen!

Thanks,

Lauren
